plugins/query-insights/src/main/java/org/opensearch/plugin/insights/rules/model/SearchQueryRecord.java

```diff
@@ -0,0 +1,216 @@
+     * The manner at which the search operation is executed.
+     */
+    public SearchType getSearchType() {
+        return searchType;
+    }
+
+    /**
+     * The search source that was executed by the query.
+     */
+    public String getSource() {
+        return source;
+    }
+
+    /**
+     * Total number of shards as part of the search query across all indices
+     */
+    public int getTotalShards() {
+        return totalShards;
+    }
+
+    /**
+     * The indices involved in the search query
+     */
+    public String[] getIndices() {
+        return indices;
+    }
+
+    /**
+     * Get the value of the query metric record
+     */
+    public T getValue() {
+        return value;
+    }
+
+    /**
+     * Set the value of the query metric record
+     */
+    public void setValue(T value) {
+        this.value = value;
+    }
+
+    /**
+     * Extra attributes and information about a search query
+     */
+    public Map<String, Object> getPropertyMap() {
+        return propertyMap;
+    }
+
+    @Override
+    public int compareTo(SearchQueryRecord<T> otherRecord) {
+        return value.compareTo(otherRecord.getValue());
+    }
+
+    public boolean equals(SearchQueryRecord<T> other) {
+        if (false == this.timestamp.equals(other.getTimestamp())
+            && this.searchType.equals(other.getSearchType())
+            && this.source.equals(other.getSource())
+            && this.totalShards == other.getTotalShards()
+            && this.indices.length == other.getIndices().length
+            && this.propertyMap.size() == other.getPropertyMap().size()
+            && this.value.equals(other.getValue())) {
+            return false;
+        }
+        for (int i = 0; i < indices.length; i++) {
+            if (!indices[i].equals(other.getIndices()[i])) {
+                return false;
+            }
+        }
+        for (String key : propertyMap.keySet()) {
+            if (!other.getPropertyMap().containsKey(key)) {
+                return false;
+            }
+            if (!propertyMap.get(key).equals(other.getPropertyMap().get(key))) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    @SuppressWarnings("unchecked")
+    private T castToValue(Object obj) throws ClassCastException {
+        try {
+            return (T) obj;
+        } catch (Exception e) {
+            log.error(String.format(Locale.ROOT, "error casting query insight record value, error: %s", e));
+            throw e;
+        }
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject();
+        builder.field(TIMESTAMP, timestamp);
+        builder.field(SEARCH_TYPE, searchType);
+        builder.field(SOURCE, source);
+        builder.field(TOTAL_SHARDS, totalShards);
+        builder.field(INDICES, indices);
+        builder.field(PROPERTY_MAP, propertyMap);
+        return builder.endObject();
```

Comment:
We should extend the XContentFragment for this class, as the toXContent is not supposed to be called directly. It can also be reused that way by starting/ending the object externally

Reply:
Refactored the code based on comment: https://github.com/opensearch-project/OpenSearch/pull/11903#discussion_r1459815439